### PR TITLE
Add `unsanitized` option to async clipboard API.

### PR DESCRIPTION
**Pull request by snianu**
Closes #????

For normative changes, the following tasks have been completed:

 * [X] Modified Web platform tests (link to pull request) https://github.com/web-platform-tests/wpt/pull/35528

Implementation commitment:

 * [ ] WebKit (https://bugs.webkit.org/show_bug.cgi?id=)
 * [X] Chromium (https://bugs.chromium.org/p/chromium/issues/detail?id=1268679)
 * [ ] Gecko (https://bugzilla.mozilla.org/show_bug.cgi?id=)

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/clipboard-apis/pull/197.html" title="Last updated on Nov 18, 2023, 1:29 AM UTC (0312e3e)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/clipboard-apis/197/ad65071...0312e3e.html" title="Last updated on Nov 18, 2023, 1:29 AM UTC (0312e3e)">Diff</a>

### CODE REVIEW

**Comment by snianu** [comment]
@evanstade Made spec changes for `unsanitized` option. Please let me know if you have any comments/concerns. Thanks!

**Comment by yisibl** [comment]
I would like the `unsanitized` option to be applied to `image/svg+xml` as well, which would solve the problem of losing `<style>` when copying SVG.  See https://bugs.chromium.org/p/chromium/issues/detail?id=1226150

File: index.bs
Comment (sanketj) [inline index.bs:808]: Should we create a section for supported MIME types (with only "text/html" in it for now) and reference it from here and the algorithm below? That will avoid replicating the same information in two places.

File: index.bs
Comment (sanketj) [inline index.bs:864]: I don't follow what is meant by ", or equal to |format| if |format| is not empty" and how it relates to the rest of the paragraph. Could you elaborate?

**Review by sanketj — CHANGES_REQUESTED** [review]

File: index.bs
Comment (snianu) [reply index.bs:864]: It basically means, the user agent may not sanitize the `|representation|'s [=representation/data=]` if `|representation|'s [=representation/MIME type=]'s [=MIME type/essence=]` is equal to `|format|.`

**Comment by snianu** [comment]
Tagging @EdgarChen in this PR as well. Could you please take a look if you are OK with this change? Thanks!

File: index.bs
Comment (snianu) [reply index.bs]: Done.

File: index.bs
Comment (sanketj) [inline index.bs:830]:
```suggestion
					1. If |format| is not in [=unsanitized mime types=], then [=reject=] |p| with {{"NotAllowedError"}} {{DOMException}} in |realm|.
```

File: index.bs
Comment (sanketj) [reply index.bs:864]: Hmm not sure I follow. I understood that the UA must not sanitize for "image/png". Could you give me an example of what you're going for with the second clause?

**Review by sanketj — CHANGES_REQUESTED** [review]

File: index.bs
Comment (snianu) [reply index.bs]: Done.

**Comment by snianu** [comment]
https://github.com/w3c/clipboard-apis/pull/197#discussion_r1378176413

Clarified in the latest iteration.

File: index.bs
Comment (sanketj) [inline index.bs:822]: Shouldn't |formats| be a dictionary instead of just a DOMString?

File: index.bs
Comment (sanketj) [inline index.bs:868]: Is this trying to say that the representation's MIME type is amongst the formats inside the sequence of `DOMString`s, which is named `unsanitized`? https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/modules/clipboard/clipboard.idl;l=1?q=clipboard.idl&sq=&ss=chromium

**Review by sanketj — CHANGES_REQUESTED** [review]

File: index.bs
Comment (snianu) [reply index.bs:822]: Yep, `|formats` is a dictionary, this variable just stores the first DOMString from that dictionary in the below step:
`Set |format| to |formats|[0].`

File: index.bs
Comment (snianu) [reply index.bs:868]: basically yes. We extract the first DOMString in the `If |formats| is not empty, then` steps and then try to match the MIME type with it in this highlighted step. In this case it will be just `text/html` mime type.

File: index.bs
Comment (evanstade) [reply index.bs:822]: but you would be iterating over `formats.unsanitized`, rather than `formats`, no?

File: index.bs
Comment (evanstade) [inline index.bs:785]: Why is this a dictionary with a single member which is an array?

If this is a dictionary simply because it makes it easier to extend with more parameters later, then I'd have expected the dictionary to be called `readOptions` or something, with an array called `unsanitizedFormats`.

File: index.bs
Comment (evanstade) [inline index.bs:826]: Feels odd to me that the IDL is written to support more than one unsanitized format, but the algorithm wording is only structured to support exactly one sanitized format. Why not iterate over the whole list and reject if any of them are not in the allowed-to-be-sanitized list?

Also if more formats were added in the future, the set of permittable formats could vary by browser. How are sites to know which formats they can ask for? Is throwing an error the best way to deal with them asking for the wrong formats to be unsanitized, or should some format values just be ignored? If throwing an error, how does the site figure out which format(s) were in error?

I don't have a ton of experience with spec language/algos. Can you find a spec mentor to help with these questions?

File: index.bs
Comment (evanstade) [inline index.bs:864]: It seems like "sanitize" should be defined.

Comment (evanstade) [reply index.bs:864]: We don't have to necessarily define the exact steps that constitute hyopthetical sanitization, but it does seem that it would be useful to describe what it's not, because another way to word this text is: "The user agent MUST NOT sanitize the data if it satisfies the below conditions". So what exactly is the user agent not allowed to do, or must it do, if it's providing "unsanitized" data?

In particular, this came up in [code review](https://chromium-review.googlesource.com/c/chromium/src/+/4779949) where, after some discussion, it was decided that the CFHTML fragment tags should be left in place. The spec should be clear about this so other vendors don't come to a different conclusion at the implementation step.

File: index.bs
Comment (snianu) [reply index.bs:785]: If we just have `sequence` with `DOMString`, then we cannot make it optional (at least the bindings generator fails in Chromium). If we make it non-optional, then it would break sites that do not want to read unsanitized formats. Dictionary helps us to make this arg optional and `unsanitized` makes it explicit that authors are opting into reading unsanitized formats i.e. when sites call `navigator.clipboard.read({unsanitized: ['text/html']})`, it is clear to the reader that the site is trying to read unsanitized `text/html` format. We got [this feedback ](https://github.com/w3c/editing/issues/315#issue-942457383)from Chrome's security team.

File: index.bs
Comment (snianu) [reply index.bs:822]: Ah you're right. Missed that part.

File: index.bs
Comment (snianu) [reply index.bs]: Addressed these concerns in the latest iteration. Please let me know if I missed anything. Thanks!

File: index.bs
Comment (snianu) [reply index.bs]: Done.

File: index.bs
Comment (sanketj) [inline index.bs:548]: What do you mean by "correct description for paste events"?

File: index.bs
Comment (sanketj) [inline index.bs:553]: Is it right to say "must not be sanitized"? We should only not sanitize if these types are passed as part of the unsanitized option.

File: index.bs
Comment (sanketj) [inline index.bs:819]:
```suggestion
		{{ClipboardUnsanitizedFormats/unsanitized}} is a [=sequence=] of {{DOMString}}s corresponding to the [=representation/mime type=] that should be handled as [=unsanitized data types=].
```

File: index.bs
Comment (sanketj) [inline index.bs:866]: Does this step need to be done for each format (ie. inside the for each loop below)?

File: index.bs
Comment (sanketj) [inline index.bs:880]: Should we include this condition in the definition of "unsanitized data types" above? Something like: Unsanitized data types are "image/png" and those that are passed by the user as part of the unsanitized option.

**Review by sanketj — CHANGES_REQUESTED** [review]

File: index.bs
Comment (snianu) [reply index.bs:553]: The intention is to not sanitize any formats that are in the `Unsanitized data types` list. If the web authors specify any other MIME types, then we will reject that MIME type during `read`.

File: index.bs
Comment (snianu) [reply index.bs:819]: `Unsanitized data types` is a list of MIME types that must not be sanitized. It doesn't describe any processing steps. If the web authors specify these MIME types in the `{{ClipboardUnsanitizedFormats/unsanitized}}`, then we will handle that according to the steps in `read` method.

File: index.bs
Comment (snianu) [reply index.bs:866]: Ah nice catch. Need to reset it at the end of the loop.

File: index.bs
Comment (snianu) [reply index.bs:880]: Yeah, this is something that I wasn't sure about. I don't know the spec language for it, but I wanted to write something that allows MIME types in `unsanitized data types` list to be part of `{{ClipboardUnsanitizedFormats/unsanitized}}`, but some MIME types(like `image/png`) should be treated like `unsanitized data types` and doesn't have to be in `{{ClipboardUnsanitizedFormats/unsanitized}}`.

Currently Webkit and Chromium allow unsanitized `image/png` read/write by-default so web authors don't have to specify it in the new `{{ClipboardUnsanitizedFormats/unsanitized}}` option during `read()`.

File: index.bs
Comment (snianu) [reply index.bs:548]: hm I guess we don't expose any MIME type that isn't supported by DataTransfer APIs or the default paste event, so I think we can remove it.

File: index.bs
Comment (sanketj) [inline index.bs:550]: Is it right to say in response to "copy and cut events" given that this doesn't apply to the `DataTransfer` or `execCommand` APIs?

Comment (snianu) [reply index.bs]: Removed.

File: index.bs
Comment (snianu) [reply index.bs]: Clarified few things.

File: index.bs
Comment (sanketj) [inline index.bs:811]: nit: You use "MUST not" here vs. "MUST NOT" above. Can we make these consistent?

Comment (sanketj) [inline index.bs:811]: Can we move the definition of "optional unsanitized data types" up below the definition of "unsanitized data types"? That would be better for readability.

File: index.bs
Comment (sanketj) [inline index.bs:832]: What's the reason this needs to be done in a separate for each loop here vs. inside the loop on line 866?

**Review by sanketj — CHANGES_REQUESTED** [review]

File: index.bs
Comment (snianu) [reply index.bs:549]: Done.

File: index.bs
Comment (snianu) [reply index.bs:832]: In the Chromium's implementation we don't post the async task to read the data from the system clipboard for formats that aren't in `unsanitized` list if this check fails when `read()` is called. That makes sense as reading those formats are waste of resources if we end up throwing an exception anyway. This is why I had to move it outside the step that posts parallel tasks to do work.

File: index.bs
Comment (snianu) [reply index.bs]: Done.

Comment (snianu) [reply index.bs]: Done.

File: index.bs
Comment (sanketj) [inline index.bs:552]: This comment still doesn't really make it clear to me why we don't sanitize images. The statement "preserve meta data" is far too high level. Can we add more details here, or even better, an example?

File: index.bs
Comment (sanketj) [inline index.bs:815]:
```suggestion
		{{ClipboardUnsanitizedFormats/unsanitized}} is a [=sequence=] of {{DOMString}}s corresponding to the [=representation/mime type=] that the author wants to be treated as [=optional unsanitized data types=].
```

File: index.bs
Comment (sanketj) [reply index.bs:835]: Ok, this sounds fine. It looks like we can bypass things like checking for permissions.

File: index.bs
Comment (sanketj) [inline index.bs:879]: Don't we also need to be include "optional unsanitized data types" here? The UA must not sanitize those either.

File: index.bs
Comment (sanketj) [inline index.bs:549]: I wonder if we should include "optional unsanitized data types" in the set of data types in "unsanitized data types"? To my comment in the algorithm, this would make it clearer that UAs must treat optional and non-optional unsanitized data types the same after the optional unsanitized data types have been validated.

File: index.bs
Comment (sanketj) [reply index.bs]: Don't think this is actually done. My ask was actually to move the definition of "optional unsanitized data types" right next to the definition of "unsanitized data types" (above or below is fine). That way, both definitions are in the same place and easier to process for the reader. Sound reasonable?

**Review by sanketj — CHANGES_REQUESTED** [review]

File: index.bs
Comment (snianu) [reply index.bs:552]: This was added as part of shipping the `image/png` MIME type in Chromium. I'm not very familiar with what metadata was getting stripped out due to sanitization, so I'll have to talk to the owner to understand all the details.

File: index.bs
Comment (snianu) [reply index.bs:879]: Ack. Added per the above comment.

File: index.bs
Comment (snianu) [reply index.bs:815]: Done.

File: index.bs
Comment (sanketj) [reply index.bs:554]: In this case, I would just leave out the note entirely and just leave it as "image/png". It would still be worth understanding this aspect better and then adding some text in the spec as explanation as a follow up.

File: index.bs
Comment (snianu) [reply index.bs]: Ack.

Comment (snianu) [reply index.bs]: Done. PTAL.

File: index.bs
Comment (snianu) [reply index.bs]: Done.

**Review by sanketj — APPROVED** [review]

**Comment by snianu** [comment]
@evanstade I think the latest iteration addresses all your concerns, but please let me know if there is anything missing. I would like to restart the I2S process as I'm not receiving any response from Mozilla folks.

**Comment by snianu** [comment]
Landing this PR. Please let me know if you have any comments and I'll address it in separate PRs.

**Comment by snianu** [comment]
@zcorpan Sorry I created a new bug in Bugzilla: https://bugzilla.mozilla.org/show_bug.cgi?id=1867331
Here is the request for position: https://chromestatus.com/feature/5192271976988672?gate=5118722541092864
For Webkit they [weren't interested](https://github.com/w3c/clipboard-apis/issues/150#issuecomment-974236367) in this feature as their sanitization process in async clipboard APIs aligns with the DataTransfer APIs.

**Comment by saschanaz** [comment]
"Implementation commitment" doesn't sound like just filing bugs but more about implementation interest 🤔

**Comment by saschanaz** [comment]
That's nice, but my understanding of commitment is that there at least should be some interest. Having a bug does not mean an interest, right?

**Comment by saschanaz** [comment]
Sure. What I feel off here is the checkbox for Gecko is filled, when there's no public implementation commitment from Gecko.

**Comment by snianu** [comment]
@saschanaz We can remove it from the spec if we only have one implementor interest, but waiting indefinitely for a signal is probably not the right way to decide whether there is interest from a particular browser. If the official stance of Firefox is negative, then please let me know and I can revert this PR and create another PR and mark it as BLOCKED. That way if Firefox or Webkit changes their stance and decide to support this feature in the future, we can unblock the PR and merge it.

**Comment by saschanaz** [comment]
I see your point, but then how are we going to track which part doesn't have implementer consensus and which part does? Having an open PR is easier for that and doing so automatically prevents any part of the spec without consensus accidentally becomes CR.

**Comment by snianu** [comment]
Note that the IDL and the language in the algorithms already allow non-supporting implementors to be in compliant with the spec while also allowing implementors who are interested in the new feature. I don't see any parts in the PR that is contentious unless I'm missing something?

**Comment by saschanaz** [comment]
I'm talking about the general process rather than the PR itself. And also I don't think we'll want to add any kind of optional features just because it's okay to not implement it.

**Comment by snianu** [comment]
> And also I don't think we'll want to add any kind of optional features without consensus

We [spent over two years trying to get consensus](https://github.com/w3c/clipboard-apis/issues/150) on this API, but we couldn't. Safari has [clearly stated their position](https://github.com/w3c/clipboard-apis/issues/150#issuecomment-974594001), so since you commented that you are not in favor of an optional parameter, should I mark Firefox's position as negative?

**Comment by saschanaz** [comment]
Our position will be announced in our standards-position issue probably by Edgar. I definitely see your frustration, but still I think merging PRs without consensus makes it hard to make sure every part of the spec has consensus (which is needed for recommendation track), and I see no benefits to merge it early.

**Comment by snianu** [comment]
@saschanaz Let's discuss this in the EditingWG meeting which is going to happen next Thursday. Without any clear stance on this API, it's hard to assume whether we have consensus or not. If you can officially state your position before that, we can definitely revert the PR, else, we will have to decide what to do with it in the EditingWG meeting where hopefully someone from Firefox can give us clear official stance on this issue. Thanks!

**Pull request merged**
